### PR TITLE
[GOBBLIN-642] Implement KafkaAvroEventKeyValueReporter

### PR DESCRIPTION
**Pull request by arjun4084346**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below! @sv2000 @htran1 please review

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-XXX

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
implement KafkaAvroEventKeyValueReporter to send kafka messages with a key using KafkaKeyValueProducerPusher

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
KafkaAvroEventKeyValueReporterTest

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/KafkaReportingFormats.java [hunk 1]
```diff
@@ -68,6 +70,10 @@
     switch (this) {
       case AVRO:
         KafkaAvroEventReporter.Builder<?> builder = KafkaAvroEventReporter.Factory.forContext(context);
+        if (properties.getProperty(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY,
+                PusherUtils.DEFAULT_KAFKA_PUSHER_CLASS_NAME).equals(PusherUtils.KAFKA_KEY_VALUE_PUSHER_CLASS_NAME)) {
+          builder = new KafkaAvroEventKeyValueReporter.BuilderImpl(context, properties);
```
Comment (sv2000) [inline gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/KafkaReportingFormats.java]: Why not KafkaAvroEventKeyValueReporter.Factory.forContext() to be consistent with KafkaAvroEventReporter?

File: gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/KafkaReportingFormats.java [hunk 2]
```diff
@@ -68,6 +70,10 @@
     switch (this) {
       case AVRO:
         KafkaAvroEventReporter.Builder<?> builder = KafkaAvroEventReporter.Factory.forContext(context);
+        if (properties.getProperty(PusherUtils.KAFKA_PUSHER_CLASS_NAME_KEY,
```
Comment (htran1) [inline gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/KafkaReportingFormats.java]: Maybe you can remove the dependency on the pusher name by instantiating through a factory name. Something like. ```Class myClass = Class.forName(<factory name>);
        Method method;
        try {
          method = myClass.getDeclaredMethod("forContext", MetricContext.class);
          builder = method.invoke(context);
        } catch (NoSuchMethodException e) {
          try {
            method = myClass.getDeclaredMethod("forContext", MetricContext.class, Properties.class);
            builder = method.invoke(context, properties);
          } catch (NoSuchMethodException e) {
            throw RuntimeException();
          }
        }```

File: gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/kafka/KafkaAvroEventKeyValueReporter.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Queue;
+
+/**
+ * {@link org.apache.gobblin.metrics.reporter.EventReporter} that emits events to Kafka as serialized Avro records with a key.
+ * Key for these kafka messages is obtained from values of properties provided via {@link #keyPropertyName}.
+ * If the GobblinTrackingEvent does not contain any of the required property, key is set to null. In that case, this reporter
+ * will act like a {@link org.apache.gobblin.metrics.kafka.KafkaAvroEventReporter}
+ */
+@Slf4j
+public class KafkaAvroEventKeyValueReporter extends KafkaAvroEventReporter {
```
Comment (htran1) [inline gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/kafka/KafkaAvroEventKeyValueReporter.java]: If you want to reuse this logic for other reporter types like KafkaJsonEventKeyValueReporter then this should extend the KafkaEventReporter.

File: gobblin-modules/gobblin-kafka-common/src/test/java/org/apache/gobblin/metrics/reporter/KafkaAvroEventKeyValueReporterTest.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    event.setName(eventName);
+    event.setNamespace(namespace);
+    metadata = Maps.newHashMap();
+    metadata.put("k1", "v1");
+    metadata.put("k2", "v2");
+    metadata.put("k3", "v3");
+    event.setMetadata(metadata);
+    context.submitEvent(event);
+
+    try {
```
Comment (htran1) [inline gobblin-modules/gobblin-kafka-common/src/test/java/org/apache/gobblin/metrics/reporter/KafkaAvroEventKeyValueReporterTest.java]: Why are the sleeps required?

Comment (arjun4084346) [reply gobblin-modules/gobblin-kafka-common/src/test/java/org/apache/gobblin/metrics/reporter/KafkaAvroEventKeyValueReporterTest.java]: Not sure, but KafkaEventReporterTest also have sleep, so I guess event submission and reporting might need some time...

Comment (htran1) [reply gobblin-modules/gobblin-kafka-common/src/test/java/org/apache/gobblin/metrics/reporter/KafkaAvroEventKeyValueReporterTest.java]: That could make this flaky. Can you remove the sleeps and see what happens? I don't think there are any other threads involved here, so I'm not sure the sleeps are required.

Comment (arjun4084346) [reply gobblin-modules/gobblin-kafka-common/src/test/java/org/apache/gobblin/metrics/reporter/KafkaAvroEventKeyValueReporterTest.java]: removed

**Pull request closed without merging**